### ISSUE TITLE
Failed building wheel

### ISSUE THREAD

**Issue by shadetree01010100** [zeromq/pyzmq#1513]
Hello, all. Installation on Ubuntu is failing:

```
pip install pyzmq~=14.7
Defaulting to user installation because normal site-packages is not writeable
Collecting pyzmq~=14.7
  Using cached pyzmq-14.7.0.zip (1.5 MB)
Building wheels for collected packages: pyzmq
  Building wheel for pyzmq (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-om77zgsk/pyzmq_2c411548d1114160a7454cde344f01f9/setup.py'"'"'; __file__='"'"'/tmp/pip-install-om77zgsk/pyzmq_2c411548d1114160a7454cde344f01f9/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-935o4ssy
       cwd: /tmp/pip-install-om77zgsk/pyzmq_2c411548d1114160a7454cde344f01f9/
  Complete output (212 lines):
  running bdist_wheel
  running build
  running build_py
  Generating grammar tables from /usr/lib/python3.7/lib2to3/Grammar.txt
  Generating grammar tables from /usr/lib/python3.7/lib2to3/PatternGrammar.txt
  creating build
  creating build/lib.linux-x86_64-3.7
  creating build/lib.linux-x86_64-3.7/zmq
  copying zmq/error.py -> build/lib.linux-x86_64-3.7/zmq
  copying zmq/__init__.py -> build/lib.linux-x86_64-3.7/zmq
  creating build/lib.linux-x86_64-3.7/zmq/sugar
  copying zmq/sugar/tracker.py -> build/lib.linux-x86_64-3.7/zmq/sugar
  copying zmq/sugar/poll.py -> build/lib.linux-x86_64-3.7/zmq/sugar
  copying zmq/sugar/attrsettr.py -> build/lib.linux-x86_64-3.7/zmq/sugar
  copying zmq/sugar/context.py -> build/lib.linux-x86_64-3.7/zmq/sugar
  copying zmq/sugar/frame.py -> build/lib.linux-x86_64-3.7/zmq/sugar
  copying zmq/sugar/__init__.py -> build/lib.linux-x86_64-3.7/zmq/sugar
  copying zmq/sugar/socket.py -> build/lib.linux-x86_64-3.7/zmq/sugar
  copying zmq/sugar/constants.py -> build/lib.linux-x86_64-3.7/zmq/sugar
  copying zmq/sugar/version.py -> build/lib.linux-x86_64-3.7/zmq/sugar
  creating build/lib.linux-x86_64-3.7/zmq/devices
  copying zmq/devices/proxydevice.py -> build/lib.linux-x86_64-3.7/zmq/devices
  copying zmq/devices/basedevice.py -> build/lib.linux-x86_64-3.7/zmq/devices
  copying zmq/devices/monitoredqueuedevice.py -> build/lib.linux-x86_64-3.7/zmq/devices
  copying zmq/devices/__init__.py -> build/lib.linux-x86_64-3.7/zmq/devices
  copying zmq/devices/monitoredqueue.py -> build/lib.linux-x86_64-3.7/zmq/devices
  creating build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_device.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_context.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_z85.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_monitor.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_constants.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_poll.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_version.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_ioloop.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_stopwatch.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_etc.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_error.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_pubsub.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/__init__.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_imports.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_reqrep.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_security.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_cffi_backend.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_log.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_auth.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_multipart.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_message.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_pair.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_monqueue.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_win32_shim.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_socket.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_retry_eintr.py -> build/lib.linux-x86_64-3.7/zmq/tests
  copying zmq/tests/test_zmqstream.py -> build/lib.linux-x86_64-3.7/zmq/tests
  creating build/lib.linux-x86_64-3.7/zmq/green
  copying zmq/green/poll.py -> build/lib.linux-x86_64-3.7/zmq/green
  copying zmq/green/core.py -> build/lib.linux-x86_64-3.7/zmq/green
  copying zmq/green/__init__.py -> build/lib.linux-x86_64-3.7/zmq/green
  copying zmq/green/device.py -> build/lib.linux-x86_64-3.7/zmq/green
  creating build/lib.linux-x86_64-3.7/zmq/green/eventloop
  copying zmq/green/eventloop/ioloop.py -> build/lib.linux-x86_64-3.7/zmq/green/eventloop
  copying zmq/green/eventloop/zmqstream.py -> build/lib.linux-x86_64-3.7/zmq/green/eventloop
  copying zmq/green/eventloop/__init__.py -> build/lib.linux-x86_64-3.7/zmq/green/eventloop
  creating build/lib.linux-x86_64-3.7/zmq/ssh
  copying zmq/ssh/tunnel.py -> build/lib.linux-x86_64-3.7/zmq/ssh
  copying zmq/ssh/__init__.py -> build/lib.linux-x86_64-3.7/zmq/ssh
  copying zmq/ssh/forward.py -> build/lib.linux-x86_64-3.7/zmq/ssh
  creating build/lib.linux-x86_64-3.7/zmq/eventloop
  copying zmq/eventloop/ioloop.py -> build/lib.linux-x86_64-3.7/zmq/eventloop
  copying zmq/eventloop/zmqstream.py -> build/lib.linux-x86_64-3.7/zmq/eventloop
  copying zmq/eventloop/__init__.py -> build/lib.linux-x86_64-3.7/zmq/eventloop
  creating build/lib.linux-x86_64-3.7/zmq/eventloop/minitornado
  copying zmq/eventloop/minitornado/ioloop.py -> build/lib.linux-x86_64-3.7/zmq/eventloop/minitornado
  copying zmq/eventloop/minitornado/util.py -> build/lib.linux-x86_64-3.7/zmq/eventloop/minitornado
  copying zmq/eventloop/minitornado/stack_context.py -> build/lib.linux-x86_64-3.7/zmq/eventloop/minitornado
  copying zmq/eventloop/minitornado/log.py -> build/lib.linux-x86_64-3.7/zmq/eventloop/minitornado
  copying zmq/eventloop/minitornado/concurrent.py -> build/lib.linux-x86_64-3.7/zmq/eventloop/minitornado
  copying zmq/eventloop/minitornado/__init__.py -> build/lib.linux-x86_64-3.7/zmq/eventloop/minitornado
  creating build/lib.linux-x86_64-3.7/zmq/eventloop/minitornado/platform
  copying zmq/eventloop/minitornado/platform/interface.py -> build/lib.linux-x86_64-3.7/zmq/eventloop/minitornado/platform
  copying zmq/eventloop/minitornado/platform/posix.py -> build/lib.linux-x86_64-3.7/zmq/eventloop/minitornado/platform
  copying zmq/eventloop/minitornado/platform/common.py -> build/lib.linux-x86_64-3.7/zmq/eventloop/minitornado/platform
  copying zmq/eventloop/minitornado/platform/__init__.py -> build/lib.linux-x86_64-3.7/zmq/eventloop/minitornado/platform
  copying zmq/eventloop/minitornado/platform/windows.py -> build/lib.linux-x86_64-3.7/zmq/eventloop/minitornado/platform
  copying zmq/eventloop/minitornado/platform/auto.py -> build/lib.linux-x86_64-3.7/zmq/eventloop/minitornado/platform
  creating build/lib.linux-x86_64-3.7/zmq/auth
  copying zmq/auth/thread.py -> build/lib.linux-x86_64-3.7/zmq/auth
  copying zmq/auth/ioloop.py -> build/lib.linux-x86_64-3.7/zmq/auth
  copying zmq/auth/base.py -> build/lib.linux-x86_64-3.7/zmq/auth
  copying zmq/auth/certs.py -> build/lib.linux-x86_64-3.7/zmq/auth
  copying zmq/auth/__init__.py -> build/lib.linux-x86_64-3.7/zmq/auth
  creating build/lib.linux-x86_64-3.7/zmq/log
  copying zmq/log/__init__.py -> build/lib.linux-x86_64-3.7/zmq/log
  copying zmq/log/handlers.py -> build/lib.linux-x86_64-3.7/zmq/log
  creating build/lib.linux-x86_64-3.7/zmq/utils
  copying zmq/utils/jsonapi.py -> build/lib.linux-x86_64-3.7/zmq/utils
  copying zmq/utils/interop.py -> build/lib.linux-x86_64-3.7/zmq/utils
  copying zmq/utils/z85.py -> build/lib.linux-x86_64-3.7/zmq/utils
  copying zmq/utils/constant_names.py -> build/lib.linux-x86_64-3.7/zmq/utils
  copying zmq/utils/sixcerpt.py -> build/lib.linux-x86_64-3.7/zmq/utils
  copying zmq/utils/garbage.py -> build/lib.linux-x86_64-3.7/zmq/utils
  copying zmq/utils/__init__.py -> build/lib.linux-x86_64-3.7/zmq/utils
  copying zmq/utils/strtypes.py -> build/lib.linux-x86_64-3.7/zmq/utils
  copying zmq/utils/win32.py -> build/lib.linux-x86_64-3.7/zmq/utils
  copying zmq/utils/monitor.py -> build/lib.linux-x86_64-3.7/zmq/utils
  creating build/lib.linux-x86_64-3.7/zmq/backend
  copying zmq/backend/select.py -> build/lib.linux-x86_64-3.7/zmq/backend
  copying zmq/backend/__init__.py -> build/lib.linux-x86_64-3.7/zmq/backend
  creating build/lib.linux-x86_64-3.7/zmq/backend/cython
  copying zmq/backend/cython/__init__.py -> build/lib.linux-x86_64-3.7/zmq/backend/cython
  creating build/lib.linux-x86_64-3.7/zmq/backend/cffi
  copying zmq/backend/cffi/context.py -> build/lib.linux-x86_64-3.7/zmq/backend/cffi
  copying zmq/backend/cffi/error.py -> build/lib.linux-x86_64-3.7/zmq/backend/cffi
  copying zmq/backend/cffi/utils.py -> build/lib.linux-x86_64-3.7/zmq/backend/cffi
  copying zmq/backend/cffi/__init__.py -> build/lib.linux-x86_64-3.7/zmq/backend/cffi
  copying zmq/backend/cffi/message.py -> build/lib.linux-x86_64-3.7/zmq/backend/cffi
  copying zmq/backend/cffi/socket.py -> build/lib.linux-x86_64-3.7/zmq/backend/cffi
  copying zmq/backend/cffi/constants.py -> build/lib.linux-x86_64-3.7/zmq/backend/cffi
  copying zmq/backend/cffi/devices.py -> build/lib.linux-x86_64-3.7/zmq/backend/cffi
  copying zmq/backend/cffi/_cffi.py -> build/lib.linux-x86_64-3.7/zmq/backend/cffi
  copying zmq/backend/cffi/_poll.py -> build/lib.linux-x86_64-3.7/zmq/backend/cffi
  copying zmq/devices/monitoredqueue.pxd -> build/lib.linux-x86_64-3.7/zmq/devices
  copying zmq/utils/buffers.pxd -> build/lib.linux-x86_64-3.7/zmq/utils
  copying zmq/utils/zmq_constants.h -> build/lib.linux-x86_64-3.7/zmq/utils
  copying zmq/utils/ipcmaxlen.h -> build/lib.linux-x86_64-3.7/zmq/utils
  copying zmq/utils/getpid_compat.h -> build/lib.linux-x86_64-3.7/zmq/utils
  copying zmq/utils/zmq_compat.h -> build/lib.linux-x86_64-3.7/zmq/utils
  copying zmq/utils/pyversion_compat.h -> build/lib.linux-x86_64-3.7/zmq/utils
  copying zmq/backend/cython/socket.pxd -> build/lib.linux-x86_64-3.7/zmq/backend/cython
  copying zmq/backend/cython/utils.pxd -> build/lib.linux-x86_64-3.7/zmq/backend/cython
  copying zmq/backend/cython/context.pxd -> build/lib.linux-x86_64-3.7/zmq/backend/cython
  copying zmq/backend/cython/message.pxd -> build/lib.linux-x86_64-3.7/zmq/backend/cython
  copying zmq/backend/cython/libzmq.pxd -> build/lib.linux-x86_64-3.7/zmq/backend/cython
  copying zmq/backend/cython/checkrc.pxd -> build/lib.linux-x86_64-3.7/zmq/backend/cython
  copying zmq/backend/cffi/_cdefs.h -> build/lib.linux-x86_64-3.7/zmq/backend/cffi
  copying zmq/backend/cffi/_verify.c -> build/lib.linux-x86_64-3.7/zmq/backend/cffi
  running build_ext
  running configure
  Using zmq-prefix /usr (found via pkg-config).
  x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include -c build/temp.linux-x86_64-3.7/scratch/check_sys_un.c -o build/temp.linux-x86_64-3.7/scratch/check_sys_un.o
  x86_64-linux-gnu-gcc -pthread build/temp.linux-x86_64-3.7/scratch/check_sys_un.o -o build/temp.linux-x86_64-3.7/scratch/check_sys_un
  ************************************************
  Configure: Autodetecting ZMQ settings...
      Custom ZMQ dir:
  creating build/temp.linux-x86_64-3.7/scratch/tmp
  cc -c /tmp/timer_createb92vkbr4.c -o build/temp.linux-x86_64-3.7/scratch/tmp/timer_createb92vkbr4.o
  /tmp/timer_createb92vkbr4.c: In function ‘main’:
  /tmp/timer_createb92vkbr4.c:2:5: warning: implicit declaration of function ‘timer_create’ [-Wimplicit-function-declaration]
       timer_create();
       ^
  cc build/temp.linux-x86_64-3.7/scratch/tmp/timer_createb92vkbr4.o -o build/temp.linux-x86_64-3.7/scratch/a.out
  build/temp.linux-x86_64-3.7/scratch/tmp/timer_createb92vkbr4.o: In function `main':
  timer_createb92vkbr4.c:(.text+0x15): undefined reference to `timer_create'
  collect2: error: ld returned 1 exit status
  x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include -Izmq/utils -Izmq/backend/cython -Izmq/devices -c build/temp.linux-x86_64-3.7/scratch/vers.c -o build/temp.linux-x86_64-3.7/scratch/vers.o
  x86_64-linux-gnu-gcc -pthread build/temp.linux-x86_64-3.7/scratch/vers.o -lzmq -lrt -o build/temp.linux-x86_64-3.7/scratch/vers
      ZMQ version detected: 4.1.4
  ************************************************
  building 'zmq.backend.cython._device' extension
  creating build/temp.linux-x86_64-3.7/zmq
  creating build/temp.linux-x86_64-3.7/zmq/backend
  creating build/temp.linux-x86_64-3.7/zmq/backend/cython
  x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DHAVE_SYS_UN_H=1 -I/usr/include -Izmq/utils -Izmq/backend/cython -Izmq/devices -I/usr/include/python3.7m -c zmq/backend/cython/_device.c -o build/temp.linux-x86_64-3.7/zmq/backend/cython/_device.o
  zmq/backend/cython/_device.c: In function ‘__Pyx_ExceptionSave’:
  zmq/backend/cython/_device.c:2456:19: error: ‘PyThreadState {aka struct _ts}’ has no member named ‘exc_type’
       *type = tstate->exc_type;
                     ^
  zmq/backend/cython/_device.c:2457:20: error: ‘PyThreadState {aka struct _ts}’ has no member named ‘exc_value’
       *value = tstate->exc_value;
                      ^
  zmq/backend/cython/_device.c:2458:17: error: ‘PyThreadState {aka struct _ts}’ has no member named ‘exc_traceback’
       *tb = tstate->exc_traceback;
                   ^
  zmq/backend/cython/_device.c: In function ‘__Pyx_ExceptionReset’:
  zmq/backend/cython/_device.c:2470:22: error: ‘PyThreadState {aka struct _ts}’ has no member named ‘exc_type’
       tmp_type = tstate->exc_type;
                        ^
  zmq/backend/cython/_device.c:2471:23: error: ‘PyThreadState {aka struct _ts}’ has no member named ‘exc_value’
       tmp_value = tstate->exc_value;
                         ^
  zmq/backend/cython/_device.c:2472:20: error: ‘PyThreadState {aka struct _ts}’ has no member named ‘exc_traceback’
       tmp_tb = tstate->exc_traceback;
                      ^
  zmq/backend/cython/_device.c:2473:11: error: ‘PyThreadState {aka struct _ts}’ has no member named ‘exc_type’
       tstate->exc_type = type;
             ^
  zmq/backend/cython/_device.c:2474:11: error: ‘PyThreadState {aka struct _ts}’ has no member named ‘exc_value’
       tstate->exc_value = value;
             ^
  zmq/backend/cython/_device.c:2475:11: error: ‘PyThreadState {aka struct _ts}’ has no member named ‘exc_traceback’
       tstate->exc_traceback = tb;
             ^
  zmq/backend/cython/_device.c: In function ‘__Pyx_GetException’:
  zmq/backend/cython/_device.c:2518:22: error: ‘PyThreadState {aka struct _ts}’ has no member named ‘exc_type’
       tmp_type = tstate->exc_type;
                        ^
  zmq/backend/cython/_device.c:2519:23: error: ‘PyThreadState {aka struct _ts}’ has no member named ‘exc_value’
       tmp_value = tstate->exc_value;
                         ^
  zmq/backend/cython/_device.c:2520:20: error: ‘PyThreadState {aka struct _ts}’ has no member named ‘exc_traceback’
       tmp_tb = tstate->exc_traceback;
                      ^
  zmq/backend/cython/_device.c:2521:11: error: ‘PyThreadState {aka struct _ts}’ has no member named ‘exc_type’
       tstate->exc_type = local_type;
             ^
  zmq/backend/cython/_device.c:2522:11: error: ‘PyThreadState {aka struct _ts}’ has no member named ‘exc_value’
       tstate->exc_value = local_value;
             ^
  zmq/backend/cython/_device.c:2523:11: error: ‘PyThreadState {aka struct _ts}’ has no member named ‘exc_traceback’
       tstate->exc_traceback = local_tb;
             ^
  error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
  ----------------------------------------
  ERROR: Failed building wheel for pyzmq

```

I've been stuck on this for what seems like an eternity. What am I missing here? Thanks!

**Comment by minrk** [comment]
The current version of pyzmq is 22.0.3. Can you try installing a more recent version? Is there a reason you are pinning pyzmq 14.7?

pyzmq 14.7 was released in 2015. Support for Python 3.7 was added in pyzmq 16 in 2018 (Python 3.7 was itself released in 2018).

**Comment by shadetree01010100** [comment]
It's a pretty tedious situation I've gotten myself into. Basically it goes like this, I've inherited a project which uses a .whl file to install. When that wheel was made, py3.5 was the name of the game, and the pinned requirement made sense. Fast forward a few years to today, and I need to spin up a copy of this project on like hardware, using Ubuntu16, which ships with py3.5. The pip 3 installed by apt (python3-pip, alongside python3-dev and python3-setuptools) fails installing another lib (cryptography-3.4.6) and states that updating pip should fix it, redirecting the user to https://pip.pypa.io/en/stable/installing/#upgrading-pip However if you do this on python 3.5, which is officially discontinued, the version of pip that gets installed cannot run in py3.5 because format strings:
```
$ python3 -m pip install -U pip
Collecting pip
  Downloading https://files.pythonhosted.org/packages/fe/ef/60d7ba03b5c442309ef42e7d69959f73aacccd0d86008362a681c4698e83/pip-21.0.1-py3-none-any.whl (1.5MB)
    100% |████████████████████████████████| 1.5MB 334kB/s 
Installing collected packages: pip
Successfully installed pip-8.1.1
You are using pip version 8.1.1, however version 21.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
$ pip --version
Traceback (most recent call last):
  File "/home/nio/.local/bin/pip", line 7, in <module>
    from pip._internal.cli.main import main
  File "/home/nio/.local/lib/python3.5/site-packages/pip/_internal/cli/main.py", line 60
    sys.stderr.write(f"ERROR: {exc}")
                                   ^
SyntaxError: invalid syntax
```
So I upgraded to python3.7, which still seems like the right choice (other than the current stable 3.9, of course), and got pip working right again (extra tedious). So now that all of that is figured out, here I am, trying to `pip install` my wheel, with its legacy dependencies, and pyzmq~=14.7. The terminal output in my OP was reproducing the issue outside of my ansible environment.

Now it turns out that the only reason I have to use *this* wheel is because of some other legacy stuff that I actually have a solution for. Given that solution, I can install a more modern wheel for this project, which doesn't even have this (pyzmq) dependency (nothing personal, of course).

I do vaguely recall dealing with this problem around the release of py3.6 and format strings, and I believe that there is an intermediate version of pip that can be installed that would also solve (some of?) my problem(s). I'm just not sure that it's worth going on that goose chase to keep running old code.

So tl;dr, I think I have the answer to my problem, as you suggested, I need to upgrade away from some of these legacy packages. Thanks!

**Comment by minrk** [comment]
Yup! That sounds tough. Upgrading Python a couple minor versions does generally mean it's a good idea to "start fresh" with the rest of your environment pinning, as lots of packages may need upgrades to work with the newer Python.

> I believe that there is an intermediate version of pip that can be installed

Yes, I believe it was pip 9 that introduced support for the `Requires-Python` metadata which is how pip decides if a package supports the target Python. Upgrading to e.g. `pip install 'pip<20'` then `pip install --upgrade pip` should get you the latest *supported* pip for your Python.
